Add Clients.withHttpHeaders() and withHttpHeader()

trustin opened the pull request:
Motivation:

There's no simple way for a user to specify HTTP headers when making a
Thrift call via *.Iface and *.AsyncIface clients.

Modifications:

- Add Clients.withHttpHeaders() and withHttpHeader() which allows a user
  to set a thread-local HTTP header manipulator function
- Make UserClient pick up the thread-local HTTP header manipulator
  function set by withHttpHeaders()
- Add SafeCloseable
- Remove RequestContext.PushHandle and use SafeCloseable instead

Result:

- A user can send a custom HTTP header without using decorator.

imasahiro commented:
> Add Client.withHttpHeaders() and withHttpHeader() which allows a user

nit: Client -> Clients

trustin commented:
@imasahiro Thanks! I've just updated the commit message and added a documentation for this feature. Please generate the Sphinx site using `./gradlew sphinx` to check what it looks like as well.

imasahiro commented:
lgtm

trustin commented:
Thank you, @imasahiro !
